queries/go.scm

```diff
@@ -35,3 +35,104 @@
 ] @string @textFragment
 
 (comment) @comment @textFragment
+
+;; What should map and list refer to in Go programs?
+;;
+;; The obvious answer is that map should refer to map and struct composite literals,
+;; and that list should refer to slice and array composite literals.
+;;
+;; There are two problems with this answer.
+;;
+;;   * The type of a composite literal is a semantic, not a syntactic property of a program.
+;;       - What is the type of T{1: 2}? It could be array, map, or slice.
+;;       - What about T{a: 1}? It could be map or struct.
+;;       - What about T{1, 2}? It could be struct, array, or slice.
+;;     Cursorless only has syntactic information available to it.
+;;
+;;   * The user might not know the type either. With a named type, the type definition might be far away.
+;;     Or it might just be offscreen. Either way, the user needs to be able to make a decision about
+;;     what scope to use using only locally available, syntactic information.
+;;     Note that this also means that has-a predicates work better than has-no predicates.
+;;     The user can locally confirm that there is a keyed element.
+;;     She cannot confirm locally that there is no keyed element; it might just not be visible.
+;;
+;; Combining all these constraints suggests the following simple rules:
+;;
+;;   * If there is a keyed element present, then it is a map.
+;;   * If there is a non-keyed element present, then it is a list.
+;;   * If there are both or neither, then it is both a map and a list.
+;;
+;; Conveniently, this is also simple to implement.
+;;
+;; This guarantees that a user always knows how to refer to any composite literal.
+;; There are cases in which being overgenerous in matching is not ideal,
+;; but they are rarer, so let's optimize for the common case.
+;; Mixed keyed and non-keyed elements are also rare in practice.
+;; The main ambiguity is with {}, but there's little we can do about that.
+;;
+;; Go users also expect that the map and list scopes will include the type definition,
+;; as well as any & before the type. (Strictly speaking it is not part of the literal,
+;; but that's not how most humans think about it.)
+;;
+;; If you are considering changing the map and list scopes, take a look at the examples in
+;; data/playground/go/maps_and_lists.go, which cover a fairly wide variety of cases.
+
+;; maps
+
+;; &T{a: 1}
+(unary_expression
+  operator: "&"
+  (composite_literal body: (literal_value (keyed_element)))
+) @map
+
+;; T{a: 1}
+(
+  (composite_literal body: (literal_value (keyed_element))) @_comp_lit
+  (#not-parent-type? @_comp_lit unary_expression)
```

Comment:
what happens if the `composite_literal` node has a parent of type `unary_expression` whose operator is not `&`?  then neither this pattern nor the other will match. Is that desirable?  If the operator is always `&` for a `unary_expression` containing a `composite_literal`, then why specify it above?

Reply:
the only valid operator is `&`. I specified it for two reasons: a form of documentation, and as future proofing in case the treesitter grammar decides that it is happy to spit out invalid unary expressions there. if you feel strongly I am happy to remove it, but I think it improves readability.